### PR TITLE
Liveness fix for struct enreg.

### PR DESCRIPTION
**Pull request by sandreenko**
Delete some unnecessary `mustInit`. The change is required for a small struct enreg change.

spmi changes:
```
benchmarks.run.windows.x64.checked.mch
Total bytes of delta: -725 (-1.10% of base)
71 total files with Code Size differences (69 improved, 2 regressed), 0 unchanged.

libraries.crossgen.windows.x64.checked
Total bytes of delta: -479 (-0.37% of base)
42 total files with Code Size differences (41 improved, 1 regressed), 0 unchanged.

libraries.crossgen2.windows.x64.checked.mch
Total bytes of delta: -3296 (-0.90% of base)
373 total files with Code Size differences (345 improved, 28 regressed), 8 unchanged.

libraries.pmi.windows.x64.checked
Total bytes of delta: -2101 (-0.62% of base)
243 total files with Code Size differences (214 improved, 29 regressed), 0 unchanged.

tests.pmi.windows.x64.checked
Total bytes of delta: -18637 (-0.81% of base)
1475 total files with Code Size differences (1472 improved, 3 regressed), 2 unchanged.

```

### CODE REVIEW

File: src/coreclr/jit/lower.cpp [hunk 1]
```diff
@@ -5915,17 +5915,35 @@ void Lowering::CheckNode(Compiler* compiler, GenTree* node)
...
+            {
+                // Emitter does not correctly handle live updates for LCL_VAR_ADDR
+                // when they are not contained, for example, `STOREIND byref(GT_LCL_VAR_ADDR not-contained)`
+                // would generate:
+                // add     r1, sp, 48   // r1 contains address of a lclVar V01.
+                // str     r0, [r1]     // a gc ref becomes live in V01, but emitter would not report it.
+                // Make sure that we use uncontained address nodes only for variables
+                // that will be marked as mustInit and will be alive throught the whole block even when tracked.
+                assert(lclVarAddr->isContained() || !varDsc->lvTracked || varTypeIsStruct(varDsc));
+                // TODO: support this assert for uses, see https://github.com/dotnet/runtime/issues/51900.
```
Comment (sandreenko) [inline src/coreclr/jit/lower.cpp:5942]: #51900

**Comment by sandreenko** [comment]
There are many failures but they all look unrelated, the same gcstress failures happened in https://dev.azure.com/dnceng/public/_build/results?buildId=1095005&view=ms.vss-test-web.build-test-results-tab, the jitstress failure presented in the previous run as well.

**Comment by sandreenko** [comment]
PTAL @BruceForstall @dotnet/jit-contrib , cc @erozenfeld

File: src/coreclr/jit/codegencommon.cpp [hunk 2]
```diff
@@ -4794,14 +4794,41 @@ void CodeGen::genCheckUseBlockInit()
...
-            /* VSW 102460 - Do not force initialization of compiler generated temps,
-                unless they are untracked GC type or structs that contain GC pointers */
-            CLANG_FORMAT_COMMENT_ANCHOR;
+            bool mustInitThisVar = false;
+
+            const bool isTemp      = varDsc->lvIsTemp;
+            const bool hasGCPtr    = varDsc->HasGCPtr();
+            const bool isTracked   = varDsc->lvTracked;
+            const bool isStruct    = varTypeIsStruct(varDsc);
+            const bool compInitMem = compiler->info.compInitMem;
```
Comment (erozenfeld) [inline src/coreclr/jit/codegencommon.cpp:4803]: All of these const declarations should be moved to before 
```
if (varDsc->lvIsTemp && !varDsc->HasGCPtr())
{
    varDsc->lvMustInit = 0;
    continue;
}
```

File: src/coreclr/jit/codegencommon.cpp [hunk 3]
```diff
@@ -4794,14 +4794,41 @@ void CodeGen::genCheckUseBlockInit()
...
 
-            if ((!varDsc->lvTracked || (varDsc->lvType == TYP_STRUCT)) && varDsc->lvOnFrame)
```
Comment (erozenfeld) [inline src/coreclr/jit/codegencommon.cpp:4802]: The new logic doesn't take into account `varDsc->lvOnFrame`. Is that is intentional?

File: src/coreclr/jit/codegencommon.cpp [hunk 4]
```diff
@@ -4794,14 +4794,41 @@ void CodeGen::genCheckUseBlockInit()
...
+                JITDUMP("must init a tracked V%02u because it a struct with a GC ref\n", varNum);
+                mustInitThisVar = true;
+            }
+            else
             {
+                // We are done with tracked or GC vars, now look at untracked vars without GC refs.
+                if (!isTracked)
+                {
+                    assert(!hasGCPtr);
+                    if (compInitMem && !isTemp)
```
Comment (BruceForstall) [inline src/coreclr/jit/codegencommon.cpp:4822]: Seems like we'll never see temps here due to `if (varDsc->lvIsTemp && !varDsc->HasGCPtr())` above?

File: src/coreclr/jit/codegencommon.cpp (hunk 3 above)
Comment (sandreenko) [reply src/coreclr/jit/codegencommon.cpp:4802]: Good question, if I add an assert like `if (mustInitThisVar) assert(varDsc->lvOnFrame);` it won't fail (x64 spmi) but it is a current implicit contract so we probably should not rely on it.
```
if (hasGCPtr && !isTracked) -> not tracked means not in register -> lvOnFrame == true
else if (hasGCPtr && isStruct) -> do not enregister structs, with https://github.com/dotnet/runtime/pull/51850/checks?check_run_id=2447290001 won't enregister structs with GC pointers -> lvOnFrame == true
if (!isTracked)  -> not tracked means not in register -> lvOnFrame == true
```

I will return the condition, thank you.

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**